### PR TITLE
New Agent's Queue Component

### PR DESCRIPTION
**Pull request by LucioDonda**
### Description

First implementation of the queue component for the 5.0 agent version

Issue #16

### CODE REVIEW

File: src/agent/CMakeLists.txt [hunk 1]
```diff
@@ -31,4 +31,5 @@ target_link_libraries(agent PRIVATE ${Boost_LIBRARIES})
 
 if(BUILD_TESTS)
     add_subdirectory(tests)
+    add_subdirectory(queue)
```
Comment (LucioDonda) [inline src/agent/CMakeLists.txt:34]: delete later

**Comment by LucioDonda** [comment]
## Tests available:

### JsonTest.
  JSONConversionComparisson
  JSONArrays
### QueueTest.
  SinglePushGetNotEmpty
  SinglePushPopEmpty
  SinglePushGetWithModule
  SinglePushPopFullWithTimeout
  MultithreadDifferentType
  MultithreadSameType
  PushMultipleSeveralSingleGets
  PushMultipleWithMessageVector
  PushVectorWithAMultipleInside
  PushMultipleGetMultiple
  PushMultipleGetMultipleWithModule
  PushSinglesleGetMultipleWithModule
  GetNextAwaitableBase
  PushAwaitable
  FifoOrderCheck

### SQLiteStorageTest.
  StoreSingleMessage
  StoreSingleMessageWithModule
  StoreMultipleMessages
  StoreMultipleMessagesWithModule
  RetrieveMessage
  RetrieveMessageWithModule
  RetrieveMultipleMessages
  RetrieveMultipleMessagesWithModule
  RemoveMessage
  RemoveMessageWithModule
  RemoveMultipleMessages
  RemoveMultipleMessagesWithModule
  GetElementCount
  GetElementCountWithModule
SQLiteStorageMultithreadedTest.
  MultithreadedStoreAndRetrieve

File: src/agent/queue/include/queue.hpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+#include <boost/asio.hpp>
+// #include <boost/asio/experimental/awaitable_operators.hpp>
+#include <cassert>
+
+#include "shared.hpp"
+#include "sqlitestorage.h"
+
+// TODO: move to a configuration setting
+constexpr int DEFAULT_MAX = 10;
```
Comment (cborla) [inline src/agent/queue/include/queue.hpp:20]: The default maximum should be a number around 10,000 elements.

File: src/agent/queue/include/sqlitestorage.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    std::unique_ptr<SQLite::Database> m_db;
+
+    /// Mutex to ensure thread-safe operations.
+    std::mutex m_mutex;
+
+    /// @brief condition variable to wait for database access
+    std::condition_variable m_cv;
+
+    // TODO: should it be atomic?
+    bool m_db_in_use = false;
```
Comment (cborla) [inline src/agent/queue/include/sqlitestorage.h:133]: Access to `m_db_in_use` is protected by a `std::mutex`, which means that concurrent access to this variable is being coordinated through a mutual exclusion mechanism. The `std::mutex` ensures that only one thread at a time can execute the critical sections that involve reading or writing `m_db_in_use`.

You can remove the TODO.

File: src/agent/queue/src/queue.cpp [hunk 4]
```diff
@@ -0,0 +1,269 @@
+#include <any>
```
Comment (cborla) [inline src/agent/queue/src/queue.cpp:1]: Already included in `queue.hpp`

File: src/agent/queue/src/queue.cpp [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+#include <chrono>
+#include <condition_variable>
```
Comment (cborla) [inline src/agent/queue/src/queue.cpp:3]: Already included in `queue.hpp`

File: src/agent/queue/src/queue.cpp [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+#include <iostream>
+#include <memory>
```
Comment (cborla) [inline src/agent/queue/src/queue.cpp:5]: Already included in `queue.hpp`

File: src/agent/queue/src/queue.cpp [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+#include <mutex>
```
Comment (cborla) [inline src/agent/queue/src/queue.cpp:6]: Already included in `queue.hpp`

File: src/agent/queue/src/queue.cpp [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+#include <queue>
```
Comment (cborla) [inline src/agent/queue/src/queue.cpp:7]: Already included in `queue.hpp`

File: src/agent/queue/src/queue.cpp [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+#include <stop_token>
+#include <unordered_map>
```
Comment (cborla) [inline src/agent/queue/src/queue.cpp:9]: Already included in `queue.hpp`

File: src/agent/queue/src/queue.cpp [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+#include <utility>
+#include <vector>
```
Comment (cborla) [inline src/agent/queue/src/queue.cpp:11]: Already included in `queue.hpp`

File: src/agent/queue/tests/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+add_executable(test_sqlitestorage
+    sqlitestorage_test.cpp
+    ../src/sqlitestorage.cpp)
+
+target_link_libraries(test_sqlitestorage
+    ${GTEST_LIBRARIES}
+    ${GTEST_MAIN_LIBRARIES}
+    SQLiteCpp
+    nlohmann_json::nlohmann_json
+    fmt::fmt)
```
Comment (cborla) [inline src/agent/queue/tests/CMakeLists.txt:36]: A new line is missing at the end

**Comment by cborla** [comment]
I think we are close to having a good version, just a few tweaks, good job.

File: src/agent/CMakeLists.txt (hunk 1 above)
Comment (jr0me) [reply src/agent/CMakeLists.txt:34]: Could we already add it with the rest of the `add_subdirectory` outside of the `BUILD_TESTS` block? And even further, already add a member in the Agent class. I've found that there were some compilation problems trying to do this.

File: src/agent/queue/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+set(CMAKE_CXX_STANDARD_REQUIRED True)
+
+set(CMAKE_CXX_FLAGS "-Wall -Wextra -Wunused -pthread")
+
+include_directories(${CMAKE_SOURCE_DIR}/include)
+
+find_package(SQLiteCpp REQUIRED)
+find_package(nlohmann_json REQUIRED)
+find_package(fmt REQUIRED)
+find_package(Boost REQUIRED COMPONENTS asio beast)
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/agent/queue/CMakeLists.txt]: no my mistake. deleted!

File: src/agent/queue/include/queue.hpp
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/agent/queue/include/queue.hpp:1]: changed to **AgentQueue**

File: src/agent/CMakeLists.txt (hunk 1 above)
Comment (LucioDonda) [reply src/agent/CMakeLists.txt]: done!

File: src/agent/queue/src/main.cpp
Comment (TomasTurina) [inline src/agent/queue/src/main.cpp:1]: I think this file should be removed as the queue will be instantiated in the Agent class: https://github.com/wazuh/wazuh-agent/blob/master/src/agent/include/agent.hpp

**Review by TomasTurina — COMMENTED** [review]
Some function and classes comments are still incomplete.

Good job!

File: src/agent/queue/src/main.cpp
Comment (LucioDonda) [reply src/agent/queue/src/main.cpp:1]: done!

**Review by cborla — APPROVED** [review]
LGTM

**Review by jr0me — APPROVED** [review]

**Review by TomasTurina — APPROVED** [review]

**Pull request merged**